wallet/types/src/wallet_tx.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serialization::{Decode, Encode};
+use std::cmp::Ordering;
+
+use common::chain::{GenBlock, Transaction};
+use common::primitives::{Id, Idable};
+
+#[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
+pub enum TxState {
+    // TxStateConfirmed, TxStateInMempool, TxStateConflicted, TxStateInactive, TxStateUnrecognized
+    /// Confirmed transaction in a block
+    #[codec(index = 0)]
+    Confirmed(Id<GenBlock>),
+    /// Unconfirmed transaction in the mempool
+    #[codec(index = 1)]
+    InMempool,
+    /// Conflicted transaction with a confirmed block
+    #[codec(index = 2)]
+    Conflicted(Id<GenBlock>),
+    /// Transaction that is not confirmed or conflicted and is not in the mempool.
+    #[codec(index = 3)]
+    Inactive,
+    /// Unrecognized state
+    #[codec(index = 4)]
+    Unrecognized,
+}
+
+#[derive(Debug, Eq, Clone, Decode, Encode)]
+pub struct WalletTx {
+    /// The actual transaction
+    tx: Transaction,
+    /// The state of this transaction
+    state: TxState,
+    /// The order of the transaction
+    order: i64,
+}
+
+impl PartialEq for WalletTx {
```

Comment:
The more I think about this, the more I realize that this is not right. Why bother and implement Ord and PartialEq for this? It's at best ambiguous. 

Why not just compare `wallet_tx.get_tx_id()` instead of specializing Eq this way?

By the same token, why not just extra the order when it's needed instead of using this as a generic ordering?

Reply:
I needed the `PartialEq` for `WalletTx` to be equal even if the order changed. For now I reverted the `PartialEq` as it was only needed for tests.

The `Ord` is needed so the transactions can be properly sorted.

Reply:
I need a little more insight from you to understand the problem. When you say "properly sorted". What do you mean? Sorting them where?

If you're sorting them in a Vec, you can use `sort_by` and provide a closure (functor, in C++ lingo) to tell it how to sort.

If you're ordering them for a BTreeMap, then use the key for the ordering.

If you're ordering them for a BTreeSet, then create a wrapper type (equivalent to functors in C++, but in a different way) that implements that sorting with deref and as_ref impls, depending on how it should be used.

Even in C++, we don't implement `WalletTx::operator<(int order1, int order2)` just for convenience. We don't do it, because it's confusing, because order means something. Types should have a meaning as well, and hence we emphasize making operations on types meaningful.

Reply:
I need to keep the transactions sorted in memory in a `BTreeMap` and the order must follow `order: Option<u64>` using a custom logic. This is to respond quickly to the RPC requests to the transaction history.

Reply:
I don't understand. Why not make the key of the map the order number?

Reply:
Because I need to query by transaction hash.

Reply:
I understand you're trying to solve such a problem.

Please see how the mempool solves this problem.

https://github.com/mintlayer/mintlayer-core/blob/master/mempool/src/pool/store.rs

There's no reason to implement an arbitrary order just for convenience. There's always another correct way. Let's chat about it.